### DISCUSSION TITLE
How to add placeholder value to header?

### DISCUSSION THREAD

**Discussion by ELymar** [ThreeMammals/Ocelot#2018] [Q&A]
## Expected Behavior / New Feature
I need to take a part of the route like `api/subscriber/{guid}` and create a header value that gets passed to the downstream services. 

I'm trying to do the following as part of `PreQueryStringBuilderMiddleware`:
```
var subscriber = context.TemplatePlaceholderNameAndValues?.FirstOrDefault(p => p.Name == "{guid}")?.Value;
context.HttpContext.Request.Headers.Add("Subscriber", new Microsoft.Extensions.Primitives.StringValues(subscriber)); 
await next.Invoke();
```

I've also tried this with no luck:
```
  "ReRoutes": [
    {
      "UpstreamPathTemplate": "/api/subscribers/{subscriberId}/widgets",
      "DownstreamScheme": "http",
      "DownstreamPathTemplate": "/api/widgets",
      "UpstreamHttpMethod": [ "Get", "Post" ],
      "ServiceName": "WidgetsUnlimited/WidgetService",
      "UpstreamHeaderTransform": {
        "SubscriberId":  "{subscriberId}"
      },
      "AuthenticationOptions": {
        "AuthenticationProviderKey": "KeyForAuth",
        "AllowedScopes": [ "openflows-watersight" ],
        "QoSOptions": {
          "TimeoutValue": 360000
        }
      }
    },
```

## Actual Behavior / Motivation for New Feature
I don't see the header in the context of the request at the downstream controller. Is there another way to achieve this?

## Specifications

  - Version: 15.0.6
  - Platform: Windows 10
  - Subsystem: ASP.NET Core

**Comment by rh101** [comment]
Try the solution described here:  https://github.com/ThreeMammals/Ocelot/issues/1267#issuecomment-649299048

**Comment by raman-m** [comment]
Why did you want to convert placeholder to a header? What's the goal?
Why not to use classic Placeholders logic to convert to query parameter and reuse it in downstream service?
```json
  "Routes": [
    {
      "UpstreamPathTemplate": "/api/subscribers/{subscriberId}/widgets",
      "DownstreamPathTemplate": "/api/widgets?subscriber={subscriberId}", // <<<--- your solution
      // redundant now
      "UpstreamHeaderTransform": {
        "SubscriberId":  "{subscriberId}"
      },
    },
```

You have to understand that by limitations our current [Header Transform Placeholders](https://ocelot.readthedocs.io/en/latest/features/headerstransformation.html#placeholders) logic supports only system predefined placeholders, but not placeholders from path templates... But Yes, it makes sense to have them in transformations too.
